### PR TITLE
Offline access experiments

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation

Implement prototype for offline enabled memeolist application that includes:

- queries
- subscriptions
- mutations

## Tasks

- [x] Test various mechanisms for caching queries 
    - [x] Http cache
    - [x] Normalized cache 
    - [x] Chaining
- [x] Test subscription on network state changes
- [x] Implement and test mutations
- [ ] Add offline support for non graphql endpoints (imgur image upload)

### CODE REVIEW

File: memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    File file = new File("/cache/");
+
+    //Size in bytes of the cache
+    int size = 1024 * 1024;
+
+    //Create the http response cache store
+    return new DiskLruHttpCacheStore(file, size);
+  }
+
+  public static SyncService getInstance(Context context) {
```
Comment (danielpassos) [inline memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java]: `getInstance()` should not receive anything if so, I prefer move `getInstance()` to a `Builder`. 

Anyway, you don't need to pass a _Context_. `MobileCore` has an `ApplicationContext` inject there on initialization, you can fire `MobileCore.getInscante().getContext()`

File: memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                     @Nonnull String webSocketUrl, Context context) {
+    DiskLruHttpCacheStore cacheStore = createHttpCache();
+
+    ApolloClient.Builder builder = ApolloClient.builder();
+    builder.serverUrl(nonNull(serverUrl, "serverUrl"))
+            .okHttpClient(nonNull(okHttpClient, "okHttpClient"))
+            .httpCache(new ApolloHttpCache(cacheStore));
+            //.subscriptionTransportFactory(new Factory(webSocketUrl, okHttpClient));
+
+    NormalizedCacheFactory cacheFactory = new LruNormalizedCacheFactory(EvictionPolicy.builder().maxSizeBytes(10 * 1024).build());
```
Comment (danielpassos) [inline memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java]: Should `maxSizeBytes` be configurable?

File: memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    //builder.normalizedCache(cacheFactory);
+    //createSQLCache(builder, context);
```
Comment (danielpassos) [inline memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java]: Why this both lines are commented?

**Comment by danielpassos** [comment]
@wtrocki In theory, for now, the only thing we need to add on our wrapper is `httpCachePolicy`?

File: memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java (hunk 1 above)
Comment (wtrocki) [reply memeolist/app/src/main/java/org/aerogear/android/app/memeolist/SyncService.java]: This is investigation PR that will be used to get improvements upstream. I'm not going to use any mobile core.

**Comment by wtrocki** [comment]
@danielpassos This is just investigation PR where I'm trying apollo cache and identifying limitations etc. After we done this, we could start thinking if/how to incorporate that into SDK.

> In theory, for now, the only thing we need to add on our wrapper is httpCachePolicy?

Http cache is not suitable and has many corner cases. Going to provide update on this and then suggest how this could be done so developers will not have limited experience.

**Comment by wtrocki** [comment]
Adjusted changes to the @darahayes server side modifications

**Comment by wtrocki** [comment]
Closing. Going to provide more comprehensive example in separate PR.

**Pull request closed without merging**